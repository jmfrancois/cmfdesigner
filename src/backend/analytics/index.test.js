const path = require('path');
const analyse = require('./index').analyse;

const TEST_ANALYTICS = analyse({ path: path.join(__dirname, './tests') });

function getAnalytics(fileName) {
	const pathToFind = path.join(__dirname, 'tests', fileName);
	return TEST_ANALYTICS.find(analytics => analytics.path === pathToFind);
}

describe('analyse', () => {
	it('should support exports', () => {
		const analytics = getAnalytics('exports.js');
		expect(analytics.export).toMatchSnapshot();
	});
	it('should support imports', () => {
		const analytics = getAnalytics('imports.js');
		expect(analytics.dependencies.length).toBe(4);
		// import defaultFromExternal, { namedFromExternal } from 'external';
		expect(analytics.dependencies[0]).toEqual({
			default: true,
			isLocal: false,
			name: 'defaultFromExternal',
			source: 'external',
		});
		expect(analytics.dependencies[1]).toEqual({
			default: false,
			isLocal: false,
			name: 'namedFromExternal',
			source: 'external',
		});
		// import defaultFromLocal, { namedFromLocal } from './local';
		expect(analytics.dependencies[2]).toMatchObject({
			default: true,
			isLocal: true,
			name: 'defaultFromLocal',
			source: './local',
		});
		expect(analytics.dependencies[3]).toMatchObject({
			default: false,
			isLocal: true,
			name: 'namedFromLocal',
			source: './local',
		});
		expect(analytics.dependencies[2].path.endsWith('/src/backend/analytics/tests/local.js')).toBe(true);
		expect(analytics.dependencies[3].path.endsWith('/src/backend/analytics/tests/local.js')).toBe(true);
	});
	it('should support components', () => {
		const analytics = getAnalytics('components.js');
		expect(analytics.components.length).toBe(2);
		expect(analytics.components[0]).toEqual({
			displayName: 'StateFullInternalComponent',
			name: 'StateFullInternalComponent',
			propTypes: true,
			type: 'class',
		});
		expect(analytics.components[1]).toEqual({
			displayName: 'PureFnInternalComponent',
			name: 'PureFnInternalComponent',
			propTypes: true,
			type: 'function',
		});
		expect(analytics.components).toMatchSnapshot();
	});
});
