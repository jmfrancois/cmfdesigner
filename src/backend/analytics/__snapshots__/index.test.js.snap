// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyse should support components 1`] = `
Array [
  Object {
    "displayName": "StateFullInternalComponent",
    "name": "StateFullInternalComponent",
    "propTypes": true,
    "type": "class",
  },
  Object {
    "displayName": "PureFnInternalComponent",
    "name": "PureFnInternalComponent",
    "propTypes": true,
    "type": "function",
  },
]
`;

exports[`analyse should support exports 1`] = `
Array [
  Object {
    "default": false,
    "path": "./local",
    "specifiers": Array [
      Object {
        "declarationType": "Identifier",
        "name": "local",
      },
    ],
  },
  Object {
    "async": false,
    "declarationType": "FunctionDeclaration",
    "default": false,
    "displayName": "function",
    "generator": false,
    "isComponent": true,
    "name": "Func",
    "params": Array [
      Object {
        "defaultValue": undefined,
        "name": "param1",
        "properties": Array [],
        "type": "Identifier",
      },
      Object {
        "defaultValue": undefined,
        "name": undefined,
        "properties": Array [
          Object {
            "name": "deconstructed",
            "shorthand": true,
          },
        ],
        "type": "ObjectPattern",
      },
      Object {
        "defaultValue": undefined,
        "name": "rest",
        "properties": Array [],
        "type": "RestElement",
      },
    ],
    "propTypes": true,
    "type": "function",
  },
  Object {
    "declarationType": "ClassDeclaration",
    "default": false,
    "displayName": "Foo",
    "isComponent": true,
    "name": "Foo",
    "propTypes": true,
    "type": "class",
  },
  Object {
    "declarationType": "VariableDeclaration",
    "declarations": Array [
      Object {
        "name": "constant",
        "value": "CONSTANT",
      },
    ],
    "default": false,
    "name": "constant",
    "type": "const",
    "value": "CONSTANT",
  },
  Object {
    "declarationType": "ObjectExpression",
    "default": true,
    "name": "undefined",
    "properties": Object {
      "Internal": Object {
        "declarationType": "Identifier",
        "key": "Internal",
        "name": "Internal",
        "type": "class",
      },
      "dep": Object {
        "declarationType": "Identifier",
        "key": "dep",
        "name": "dep",
        "type": "import",
      },
      "fnInternal": Object {
        "async": false,
        "declarationType": "Identifier",
        "generator": false,
        "key": "fnInternal",
        "name": "fnInternal",
        "params": Array [
          Object {
            "defaultValue": undefined,
            "name": "param1",
            "properties": Array [],
            "type": "Identifier",
          },
          Object {
            "defaultValue": undefined,
            "name": undefined,
            "properties": Array [
              Object {
                "name": "deconstructed",
                "shorthand": true,
              },
            ],
            "type": "ObjectPattern",
          },
          Object {
            "defaultValue": undefined,
            "name": "rest",
            "properties": Array [],
            "type": "RestElement",
          },
        ],
        "type": "function",
      },
      "fnstar": Object {
        "async": false,
        "declarationType": "Identifier",
        "generator": true,
        "key": "fnstar",
        "name": "fnstar",
        "params": Array [],
        "type": "function",
      },
      "objectMethod": Object {
        "async": false,
        "generator": false,
        "key": "objectMethod",
        "params": Array [
          Object {
            "defaultValue": undefined,
            "name": "foo",
            "properties": Array [],
            "type": "Identifier",
          },
          Object {
            "defaultValue": undefined,
            "name": undefined,
            "properties": Array [
              Object {
                "name": "bar",
                "shorthand": true,
              },
            ],
            "type": "ObjectPattern",
          },
          Object {
            "defaultValue": undefined,
            "name": "rest",
            "properties": Array [],
            "type": "RestElement",
          },
        ],
        "type": "function",
      },
    },
    "type": "object",
  },
]
`;
